platform/pax-web-jetty/src/main/java/org/ops4j/pax/web/service/jetty/internal/JettyServerWrapper.java

```diff
@@ -0,0 +1,639 @@
+    lazyLoad = lazy;
+    this.storeDirectory = directory;
+    this.sessionCookieMaxAge = maxAge;
+  }
+
+  HttpServiceContext getContext(final HttpContext httpContext) {
+    readLock.lock();
+    try {
+      ServletContextInfo servletContextInfo = contexts.get(httpContext);
+      if (servletContextInfo != null) {
+        return servletContextInfo.getHandler();
+      }
+      return null;
+    } finally {
+      readLock.unlock();
+    }
+  }
+
+  HttpServiceContext getOrCreateContext(final Model model) {
+    return getOrCreateContext(model.getContextModel());
+  }
+
+  HttpServiceContext getOrCreateContext(final ContextModel model) {
+    final HttpContext httpContext = model.getHttpContext();
+    ServletContextInfo context = null;
+    try {
+      readLock.lock();
+      if (contexts.containsKey(httpContext)) {
+        context = contexts.get(httpContext);
+        context.incrementRefCount();
+      } else {
+        try {
+          readLock.unlock();
+          writeLock.lock();
+          if (!contexts.containsKey(httpContext)) {
+            LOG.debug(
+                "Creating new ServletContextHandler for HTTP context [{}] and model [{}]",
+                httpContext,
+                model);
+
+            context = new ServletContextInfo(this.addContext(model));
+            contexts.put(httpContext, context);
+            // don't increment! - it's already == 1 after creation
+            //						context.incrementRefCount();
+          } else {
+            context = contexts.get(httpContext);
+            context.incrementRefCount();
+          }
+        } finally {
+          readLock.lock();
+          writeLock.unlock();
+        }
+      }
+    } finally {
+      readLock.unlock();
+    }
+    return context.getHandler();
+  }
+
+  void removeContext(final HttpContext httpContext, boolean force) {
+    ServletContextInfo context;
+    try {
+      readLock.lock();
+      context = contexts.get(httpContext);
+      if (context == null) {
+        return;
+      }
+      int nref = context.decrementRefCount();
+      if ((force && !(httpContext instanceof SharedWebContainerContext)) || nref <= 0) {
+        try {
+          readLock.unlock();
+          writeLock.lock();
+          LOG.debug("Removing ServletContextHandler for HTTP context [{}].", httpContext);
+          context = contexts.remove(httpContext);
+        } finally {
+          readLock.lock();
+          writeLock.unlock();
+        }
+      } else {
+        LOG.debug(
+            "ServletContextHandler for HTTP context [{}] referenced [{}] times.",
+            httpContext,
+            nref);
+        return;
+      }
+    } finally {
+      readLock.unlock();
+    }
+    // Destroy the context outside of the locking region
+    if (context != null) {
+      HttpServiceContext sch = context.getHandler();
+      sch.unregisterService();
+      try {
+        sch.stop();
+      } catch (Throwable t) { // CHECKSTYLE:SKIP
```

Comment:
❗️ Make sure to catch `VirtualMachineError` first and throw it back. (even though this is code mostly copied)